File: T-Bank/Algorithms-and-Data-Structures-2025/5sprint-Strings/contest/C-Двигай/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(|a| + |b|)
    Memory Complexity O(|a| + |b|)
    """
    a: str = sys.stdin.readline().rstrip()
    b: str = sys.stdin.readline().rstrip()
    k: int = len(b)
    if k > len(a):
        sys.stdout.write('-1\n')
        return
    q1, m1 = 100000009, 100000003
    q2, m2 = 100000007, 100000019
    hashmap: dict[int, dict[int, int]] = dict()
    prefix_sum1: list[int] = [0]
    prefix_sum2: list[int] = [0]
    h1 = h2 = 0
    for i in range(2 * k - 1):
        h1 = (h1 * q1 + ord(b[i % k])) % m1
        h2 = (h2 * q2 + ord(b[i % k])) % m2
        if i + 1 >= k:
            if h1 not in hashmap: hashmap[h1] = dict()
            if h2 not in hashmap[h1]: hashmap[h1][h2] = 0
            if i >= k:
                q_pow1: int = pow(q1, k, m1)
                q_pow2: int = pow(q2, k, m2)
                hash_code1: int = (h1 - prefix_sum1[i - k + 1] * q_pow1 % m1 + m1) % m1
                hash_code2: int = (h2 - prefix_sum2[i - k + 1] * q_pow2 % m2 + m2) % m2
                if hash_code1 not in hashmap: hashmap[hash_code1] = dict()
                if hash_code2 not in hashmap[hash_code1]: hashmap[hash_code1][hash_code2] = 0 
                hashmap[hash_code1][hash_code2] += 1
            else:
                hashmap[h1][h2] += 1
        prefix_sum1.append(h1)
        prefix_sum2.append(h2)
    output: int = 0 
    prefix_sum1 = [0]
    prefix_sum2 = [0]
    h1 = h2 = 0
    for i in range(len(a)):
        h1 = (h1 * q1 + ord(a[i])) % m1
        h2 = (h2 * q2 + ord(a[i])) % m2
        if i + 1 >= k:
            if i >= k:
                q_pow1: int = pow(q1, k, m1)
                q_pow2: int = pow(q2, k, m2)
                hash_code1: int = (h1 - prefix_sum1[i - k + 1] * q_pow1 % m1 + m1) % m1
                hash_code2: int = (h2 - prefix_sum2[i - k + 1] * q_pow2 % m2 + m2) % m2
                if hash_code1 in hashmap and hash_code2 in hashmap[hash_code1]: output += 1
            else:
                if h1 in hashmap and h2 in hashmap[h1]: output += 1
                
        prefix_sum1.append(h1)
        prefix_sum2.append(h2)
    sys.stdout.write('{}\n'.format(output))


if __name__ == '__main__':
    solution()

File: T-Bank/Algorithms-and-Data-Structures-2025/5sprint-Strings/contest/C-Двигай/solution.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;

long long modPow(long long base, long long exp, long long mod) {
    long long result = 1;
    while (exp > 0) {
        if (exp % 2 == 1) {
            result = (result * base) % mod;
        }
        base = (base * base) % mod;
        exp /= 2;
    }
    return result;
}

void solution() {
    /*
    Time Complexity O(|a| + |b|)
    Memory Complexity O(|a| + |b|)
    */
    string a, b;
    cin >> a >> b;
    int k = b.size();
    if (k > a.size()) {
        printf("0\n");
        return;
    }
    const long long q1 = 100000009;
    const long long m1 = 100000003;
    const long long q2 = 100000007;
    const long long m2 = 100000019;
    unordered_map<long long, unordered_map<long long, int>> hashmap;
    vector<long long> prefixSum1 = {0}, prefixSum2 = {0};
    long long h1 = 0, h2 = 0;
    for (int i = 0; i < k * 2 - 1; ++i) {
        h1 = (h1 * q1 + static_cast<int>(b[i % k])) % m1;
        h2 = (h2 * q2 + static_cast<int>(b[i % k])) % m2;
        if (i + 1 >= k) {
            if (i >= k) {
                long long qPow1 = modPow(q1, k, m1);
                long long qPow2 = modPow(q2, k, m2);
                long long hashCode1 = (h1 - prefixSum1[i - k + 1] * qPow1 % m1 + m1) % m1;
                long long hashCode2 = (h2 - prefixSum2[i - k + 1] * qPow2 % m2 + m2) % m2;
                ++hashmap[hashCode1][hashCode2];
            } else ++hashmap[h1][h2];
        }
        prefixSum1.push_back(h1);
        prefixSum2.push_back(h2);
    }
    long long output = 0;
    prefixSum1 = {0};
    prefixSum2 = {0};
    h1 = 0, h2 = 0;
    for (int i = 0; i < a.size(); ++i) {
        h1 = (h1 * q1 + static_cast<int>(a[i])) % m1;
        h2 = (h2 * q2 + static_cast<int>(a[i])) % m2;
        if (i + 1 >= k) {
            if (i >= k) {
                long long qPow1 = modPow(q1, k, m1);
                long long qPow2 = modPow(q2, k, m2);
                long long hashCode1 = (h1 - prefixSum1[i - k + 1] * qPow1 % m1 + m1) % m1;
                long long hashCode2 = (h2 - prefixSum2[i - k + 1] * qPow2 % m2 + m2) % m2;
                if (hashmap.count(hashCode1) && hashmap[hashCode1].count(hashCode2)) ++output;
            } else if (hashmap.count(h1) && hashmap[h1].count(h2)) ++output;
        }
        prefixSum1.push_back(h1);
        prefixSum2.push_back(h2);
    }
    printf("%lld\n", output);
}

int main() {
    solution();
}